internal/proxy/setting.go

```diff
@@ -0,0 +1,156 @@
+package proxy
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/ubuntu/decorate"
+)
+
+type protocol string
+
+const (
+	allProxy   protocol = "all"
+	noProxy    protocol = "no"
+	httpProxy  protocol = "http"
+	httpsProxy protocol = "https"
+	ftpProxy   protocol = "ftp"
+	socksProxy protocol = "socks"
+)
+
+// proxySetting represents a proxy setting to be applied on the system.
+type proxySetting struct {
+	protocol   protocol
+	escapedURL string // scheme://host:port, including escaped user:password if available, verbatim if no_proxy
+
+	url *url.URL
+}
+
+// setConfig parses and validates the given proxy settings, returning them in a
+// format ready to be applied on the system.
+func setConfig(http, https, ftp, socks, noproxy string) (proxies []proxySetting, err error) {
+	defer decorate.OnError(&err, "couldn't set proxy configuration")
+
+	if useGlobalHTTPProxy(http, https, ftp, socks) {
+		https = http
+		ftp = http
+		socks = http
+
+		setting, err := newSetting(allProxy, http)
+		if err != nil {
+			return nil, err
+		}
+		proxies = append(proxies, *setting)
+	}
+
+	if http != "" {
+		setting, err := newSetting(httpProxy, http)
+		if err != nil {
+			return nil, err
+		}
+		proxies = append(proxies, *setting)
+	}
+
+	if https != "" {
+		setting, err := newSetting(httpsProxy, https)
+		if err != nil {
+			return nil, err
+		}
+		proxies = append(proxies, *setting)
+	}
+
+	if ftp != "" {
+		setting, err := newSetting(ftpProxy, ftp)
+		if err != nil {
+			return nil, err
+		}
+		proxies = append(proxies, *setting)
+	}
+
+	if socks != "" {
+		setting, err := newSetting(socksProxy, socks)
+		if err != nil {
+			return nil, err
+		}
+		proxies = append(proxies, *setting)
+	}
+
+	if noproxy != "" {
+		setting, err := newSetting(noProxy, noproxy)
+		if err != nil {
+			return nil, err
+		}
+		proxies = append(proxies, *setting)
+	}
```

Comment:
In order to avoid this, maybe we can change the function args to be a variadic of "pairs" e.g. (protocol, url)?
Then we could also have a "global" map with the protocols as well, e.g.:
```Go
protocols := map[string]protocol {
"http": httpProxy,
...
}
```
It would require quite a few code changes to propagate this behavior up the chain, but maybe this would help avoiding having calls with too many empty strings in the process and ease our life should the number of supported protocols increase (I'm not even sure if this is possible, tbh).
This is more of a rambling and shared thinking rather than a suggestion, so let me know what you think about it.